bin/upgrade-groups.bash

```diff
@@ -0,0 +1,97 @@
+#!/bin/bash
+
+set -eu -o pipefail
+
+here="$(dirname "$(readlink -f "$0")")"
+usage() {
+    echo "COMMANDS:" 1>&2
+    echo "  <prefix> logs                              Check group upgrade logs" 1>&2
+    echo "      args: <group> [<options>]" 1>&2
+    echo "  <prefix> list-groups                    List groups " 1>&2
+    echo "      args: [<options>]" 1>&2
+    echo "  <prefix> upgrade-groups                    Run upgrade playbook " 1>&2
+    echo "      args: [<options>]" 1>&2
+    exit 1
+}
+
+
+if [ $# -lt 2 ]; then
+    usage
+else
+    export prefix="${1}"
+fi
+
+# shellcheck source=bin/common.bash
+source "${here}/common.bash"
+
+# shellcheck source=bin/common.bash
+source "${here}/inventory-parser.bash"
+
+list_groups() {
+    groups=$(readInventoryGroups ${config[groups_inventory_file]})
+    for group in $groups; do 
+        log_info "- $group"
+    done
+}
+
+upgrade_groups() {
+    local -a groups
+    for group in $(readInventoryGroups "${config[groups_inventory_file]}" | sort); do
+      if ! [[ "${group}" =~ ^(all|etcd|kube_node|.*:.*)$ ]]; then
+        groups+=("${group}")
+      fi
+    done
+
+    local -A group_lengths
+    group_lengths["kube_control_plane"]="$(getGroupHosts "${config[groups_inventory_file]}" "kube_control_plane" | wc -w)"
+    for group in "${groups[@]}"; do
+      group_lengths["${group}"]="$(getGroupHosts "${config[groups_inventory_file]}" "${group}" | wc -w)"
```

Comment:
Create a variable here to save the largest group so that you can use it for the index range later. Just to be sure we never exceed that number of nodes

```suggestion
      group_lengths["${group}"]="$(getGroupHosts "${config[groups_inventory_file]}" "${group}" | wc -w)"
      if [[ "${group_lengths["${group}"]}" -gt "${largest_group_length}" ]]; then
        largest_group_length="${group_lengths["${group}"]}"
      fi
```